### PR TITLE
Add Metadata log files

### LINKED ISSUE #8183: Add metadata log context and log files
##### Enabler description

Add [metadata log](https://github.com/netdata/docs/blob/master/architecture/agent/db/database-log-format-pros-cons.md) threads, implement log file logic in append only manner.

No clean-up or log replay possible at this point. The needed data structures in the agent that would allow those operations don't exist yet.

Introduce the backwards compatible extensions to the `plugins.D` language, but only for flushing. No code will be added to the parsing code in `plugins_d.c`. This is the same situation as what streaming does with`rrdpush.c`. Parsing always goes to `plugins_d.c` but flushing is handled by the context of the producer.

- The metadata log files are next to the dbengine instance datafiles/journalfiles.
- The metadata log depends on the corresponding instance having been set-up beforehand.

- At all host/chart/dimension creation update points in the code write the plugins.D command lines to the metadata log files.

###### Acceptance criteria

* Check approval.
* A logfile that shows metadata operations in the new plugins.d protocol.

### PR DESCRIPTION
**Pull request by mfundul**
##### Summary
Fixes #8183 
##### Component Name
database
daemon
##### Test Plan

View the contents of the metadata log files as they are appended with a command such as:
```
tail -f /var/cache/netdata/dbengine/metadatalog-00001-00001.mlf | strings
```

##### Additional Information

Add metadata log threads, implement log file logic in append only manner.

No clean-up or log replay possible at this point. The needed data structures in the agent that would allow those operations don't exist yet.

### CODE REVIEW

File: database/engine/metadata_log/logfile.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+extern int destroy_metadata_logfile(struct metadata_logfile *metalogfile);
+extern int create_metadata_logfile(struct metadata_logfile *metalogfile);
+extern int load_metadata_logfile(struct metalog_instance *ctx, struct metadata_logfile *logfile);
+extern void init_metadata_record_log(struct metalog_instance *ctx);
+extern int add_new_metadata_logfile(struct metalog_instance *ctx, unsigned tier, unsigned fileno);
+extern int init_metalog_files(struct metalog_instance *ctx);
+extern void finalize_metalog_files(struct metalog_instance *ctx);
+
+
+#endif /* NETDATA_LOGFILE_H */
```
Comment (stelfrag) [inline database/engine/metadata_log/logfile.h:77]: Complains about `no newline` at the end of file :smile:

File: database/engine/metadata_log/metadatalogapi.h [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+extern void metalog_commit_update_chart(RRDSET *st);
+extern void metalog_commit_delete_chart(RRDSET *st);
+extern void metalog_commit_update_dimension(RRDDIM *rd);
+extern void metalog_commit_delete_dimension(RRDDIM *rd);
+
+/* must call once before using anything */
+extern int metalog_init(struct rrdengine_instance *rrdeng_parent_ctx);
+extern int metalog_exit(struct metalog_instance *ctx);
+
+#endif /* NETDATA_METADATALOGAPI_H */
```
Comment (stelfrag) [inline database/engine/metadata_log/metadatalogapi.h:18]: Newline here as well :exploding_head:

File: configure.ac [hunk 3]
```diff
@@ -420,6 +420,9 @@ fi
 AC_MSG_RESULT([${enable_dbengine}])
 AM_CONDITIONAL([ENABLE_DBENGINE], [test "${enable_dbengine}" = "yes"])
 
+AC_DEFINE([ENABLE_METADATALOG], [1], [netdata metadata log])
```
Comment (amoss) [inline configure.ac:423]: Why do we have this? Won't it be switched on all the time...

Comment (mfundul) [reply configure.ac:423]: Yes, it's a leftover, it should go away.

Comment (mfundul) [reply configure.ac:423]: Wait, it's not there, it was the initial commit. You can look at the final version to make sure.

Comment (mfundul) [reply configure.ac:423]: Branch mess. Let me rebase first.

File: database/engine/metadata_log/metadatalogapi.c [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    metalog_commit_record(ctx, buffer, METALOG_COMMIT_DELETION_RECORD);
+}
+
+void metalog_commit_update_host(RRDHOST *host)
+{
+    struct metalog_instance *ctx;
+    BUFFER *buffer;
+
+    /* Metadata are only available with dbengine */
+    if (!host->rrdeng_ctx)
```
Comment (amoss) [inline database/engine/metadata_log/metadatalogapi.c:33]: The lock in L39 below guards against the read accesses to the host for the sprintf but not this or L36.

Comment (mfundul) [reply database/engine/metadata_log/metadatalogapi.c:33]: This code has been taken from streaming. The host lock is primarily protects the list of charts. Moreover, it has been use to protect the list of labels when labels were added and in some cases to protect some writeable fields in the host structure.

However, most of those things are not 100% safe as we proved when we crashed the agent. The immutable values however are never protected in the current locking scheme. One thing that never changes is exactly what is in L36 which is `host->rrdeng_ctx->metalog_ctx`.

This is a way too big discussion to have but I would argue that protecting L36 and many more like it does not give any protection with the way things are currently.

File: database/engine/metadata_log/metadatalogapi.c [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+                   default_rrd_update_every,
+                   host->os,
+                   host->timezone,
+                   (host->tags) ? host->tags : "");
+
+    netdata_rwlock_rdlock(&host->labels_rwlock);
+    struct label *labels = host->labels;
+    while (labels) {
+        buffer_sprintf(buffer
+            , "LABEL \"%s\" = %d %s\n"
```
Comment (amoss) [inline database/engine/metadata_log/metadatalogapi.c:57]: The `%s` should be quoted as `"%s"`.

Comment (amoss) [reply database/engine/metadata_log/metadatalogapi.c:57]: I don't think that you need the `label_source`

File: database/engine/metadata_log/metadatalogapi.c [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+            , labels->key
+            , (int)labels->label_source
+            , labels->value);
+
+        labels = labels->next;
+    }
+    netdata_rwlock_unlock(&host->labels_rwlock);
+
+    buffer_strcat(buffer, "OVERWRITE labels\n");
```
Comment (amoss) [inline database/engine/metadata_log/metadatalogapi.c:66]: As a note - I don't think we are required to match the streaming implementation of this.

File: database/engine/metadata_log/metadatalog.c [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+    mlf_flush_records_buffer(wc);
+}
+
+#define MAX_CMD_BATCH_SIZE (256)
+
+void metalog_worker(void* arg)
+{
+    struct metalog_worker_config *wc = arg;
+    struct metalog_instance *ctx = wc->ctx;
+    uv_loop_t* loop;
```
Comment (amoss) [inline database/engine/metadata_log/metadatalog.c:229]: Why the heap instead of the stack?

File: database/engine/metadata_log/metadatalogapi.c [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+ * Returns 0 on success, negative on error
+ */
+int metalog_init(struct rrdengine_instance *rrdeng_parent_ctx)
+{
+    struct metalog_instance *ctx;
+    int error;
+
+    rrdeng_parent_ctx->metalog_ctx = ctx = callocz(1, sizeof(*ctx));
+
+    ctx->max_disk_space = 32 * 1048576LLU;
```
Comment (amoss) [inline database/engine/metadata_log/metadatalogapi.c:240]: This should be configurable

File: database/engine/metadata_log/metadatalogapi.c (hunk 5 above)
Comment (mfundul) [reply database/engine/metadata_log/metadatalogapi.c:57]: This is copy-paste from streaming for now. We will fix the `LABEL` command in https://github.com/netdata/netdata/issues/8181 as you're saying.

File: database/engine/metadata_log/metadatalogapi.c (hunk 6 above)
Comment (mfundul) [reply database/engine/metadata_log/metadatalogapi.c:66]: Yes, you are right, we will fix this in https://github.com/netdata/netdata/issues/8181

File: database/engine/metadata_log/metadatalogapi.c (hunk 8 above)
Comment (mfundul) [reply database/engine/metadata_log/metadatalogapi.c:240]: We will implement this during compaction.

File: database/engine/metadata_log/metadatalog.c (hunk 7 above)
Comment (mfundul) [reply database/engine/metadata_log/metadatalog.c:229]: This is just following the best practices of `test/` code in libuv repo. There is no real reason behind this.

File: database/engine/metadata_log/metadatalogprotocol.h [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+#define METALOG_OTHER               (3) /* reserved */
+
+/*
+ * Metadata log record header
+ */
+struct rrdeng_metalog_record_header {
+    /* when set to STORE_PADDING jump to start of next block */
+    uint8_t type;
+
+    uint32_t reserved; /* reserved for future use */
```
Comment (amoss) [inline database/engine/metadata_log/metadatalogprotocol.h:38]: This doesn't work well for either forward or backward compatibility, change it to a header_len field instead.

Comment (mfundul) [reply database/engine/metadata_log/metadatalogprotocol.h:38]: I'm not keeping those yeah. We'll have to see a little bit later how I'll shape this, I'm not sure what I'm going to need yet.

File: database/engine/metadata_log/logfile.c [hunk 10]
```diff
@@ -0,0 +1,482 @@
...
+    unsigned pos, size;
+    struct metadata_logfile *metalogfile;
+
+    if (unlikely(NULL == ctx->records_log.buf || 0 == ctx->records_log.buf_pos)) {
+        return;
+    }
+    /* care with outstanding records when switching metadata log files */
+    metalogfile = ctx->metadata_logfiles.last;
+
+    io_descr = mallocz(sizeof(*io_descr));
```
Comment (amoss) [inline database/engine/metadata_log/logfile.c:52]: Does this leak? (I can't find the free)

Comment (amoss) [reply database/engine/metadata_log/logfile.c:52]: Ah no - there it is in the callback :)

**Comment by amoss** [comment]
This is inside the log:
```
HOST "7afcb008-99d6-11ea-bd72-0242ac110002" "fba9008562ca" "fba9008562ca" 1 "linux" "Etc/UTC" ""LABEL "_os_name" = 0 unknown
```

The extra quote looks wrong: `""LABEL`

**Comment by mfundul** [comment]
> 
> 
> This is inside the log:
> 
> ```
> HOST "7afcb008-99d6-11ea-bd72-0242ac110002" "fba9008562ca" "fba9008562ca" 1 "linux" "Etc/UTC" ""LABEL "_os_name" = 0 unknown
> ```
> 
> The extra quote looks wrong: `""LABEL`

You're right. Just added a new line character.

**Pull request merged**